src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -124,11 +126,35 @@ public async Task RunApplicationAsync(CancellationToken cancellationToken = defa
                 await lifecycleHook.AfterResourcesCreatedAsync(_model, cancellationToken).ConfigureAwait(false);
             }
         }
+        catch
+        {
+            _shutdownCts.Cancel();
+            throw;
+        }
         finally
         {
             AspireEventSource.Instance.DcpModelCreationStop();
         }
     }
+
+    public async Task StopAsync(CancellationToken cancellationToken)
+    {
+        _shutdownCts.Cancel();
+        var shutdownTasks = new List<Task>();
+        if (_resourceWatchTask is { } resourceTask)
+        {
+            shutdownTasks.Add(resourceTask);
+        }
+
+        foreach (var (_, (cts, logTask)) in _logStreams)
+        {
+            cts.Cancel();
+            shutdownTasks.Add(logTask);
+        }
+
+        await Task.WhenAll(shutdownTasks).WaitAsync(cancellationToken).ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
```

Comment:
Should errors here be logged? (ignoring OCE)

IMO try/catch with log would help prevent errors from being ignored.